stl/inc/ranges

```diff
@@ -22,15 +25,461 @@ _STL_DISABLE_CLANG_WARNINGS
+                    _Construct_in_place(_Val, _STD move(_That._Val));
+                    _Engaged = true;
+                } else {
+                    ; // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept(is_nothrow_move_constructible_v<_Ty>) {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_That._Engaged) {
+                _Construct_in_place(_Val, _STD move(_That._Val));
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return _Engaged;
+        }
+
+        constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(*this);
+            return _Val;
+        }
+        constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(*this);
+            return _Val;
+        }
+
+    private:
+        union {
+            _Nontrivial_dummy_type _Dummy;
+            _Ty _Val;
+        };
+        bool _Engaged;
+    };
+
+    // clang-format off
+    template <_Copy_constructible_object _Ty>
+        requires constructible_from<_Ty>
+            && (copyable<_Ty>
+                || (is_nothrow_copy_constructible_v<_Ty>
+                    && (movable<_Ty> || is_nothrow_move_constructible_v<_Ty>)))
+    class _Semiregular_box<_Ty> {
+        // clang-format on
+    public:
+        _Semiregular_box() = default;
+
+        template <class... _Types>
+        constexpr _Semiregular_box(in_place_t, _Types&&... _Args) noexcept(
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            : _Val(_STD forward<_Types>(_Args)...) {}
+
+        _Semiregular_box(const _Semiregular_box&) = default;
+        _Semiregular_box(_Semiregular_box&&)      = default;
+        _Semiregular_box& operator=(const _Semiregular_box&) requires copyable<_Ty> = default;
+        _Semiregular_box& operator=(_Semiregular_box&&) requires movable<_Ty> = default;
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _That._Val);
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _STD move(_That._Val));
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return true;
+        }
+
+        constexpr _Ty& operator*() noexcept {
+            return _Val;
+        }
+        constexpr const _Ty& operator*() const noexcept {
+            return _Val;
+        }
+
+    private:
+        /* [[no_unique_address]] */ _Ty _Val = _Ty();
+    };
+#else // ^^^ no workaround / workaround vvv // FIXME: these are ABI incompatible

```

Comment:
There's still a "FIXME" here.

Background: The exposition-only class template _`semiregular-box`_ behaves as if it were `std::optional` with a couple of modifications. The implementation below implements that behavior directly by layering behaviors onto `std::optional`. The implementation above takes advantage of WG21-P0848 to implement the equivalent of `std::optional` with 500 fewer lines of boilerplate and 6 fewer levels of inheritance; while I haven't actually verified it improves throughput I can't imagine how it would not. 

In the process of writing the P0848 version I discovered a bug in MSVC's very recent (not yet released) implementation of that feature, so there are two implementations here. The bug will probably be fixed for 16.8 - I verified with the compiler developer that it's an easily fixed "oops I forgot a case" bug and not "I'll need weeks to refactor the compiler" - so I'd like to keep both implementations here and covered by tests.

Of course they're not strictly ABI compatible, which could result in conflicts linking MSVC-compiled TUs that use these views with clang-compiled TUs that use them. I think it would be preferable to use the workaround unconditionally but keep the other implementation right here where I can manually enable it for test runs.